python/rmm/_lib/device_buffer.pyx

```diff
@@ -57,17 +59,36 @@ cdef class DeviceBuffer:
         buf.c_obj = move(ptr)
         return buf
 
+    @staticmethod
+    @cython.boundscheck(False)
+    cdef DeviceBuffer c_frombytes(const unsigned char[::1] b,
```

Comment:
Does this handle `memoryview` objects in general? If so it would be nice if we could have an overloaded non-const version to handle things like `bytearray` or other non-readonly objects.

Reply:
It handles `memoryview` objects as long as they are 1-D, contiguous, and of type `unsigned char`.

The `const` just means that we won't modify the underlying buffer (much like if we took `const char*` and were given `char*`). So any mutable object meeting the constraints above could be passed in as well. This just promises we won't modify the underlying data. IOW `bytearray` is permissible and we test for this below.

As to other memory types and layouts that don't meet these requirements, users could coerce them into something that meets these requirements (with a minimal amount of copying to ensure contiguous data). Decided not to do that for the user as I wanted them to know we are not preserving the type or layout of their data. Though they could take the resulting `DeviceBuffer` and layer this information back on top by constructing some other array (like CuPy or Numba) and adding this info to that object.

Reply:
Agreed with not implicitly making things contiguous at this level. I'm surprised that a bytearray is allowed for a `const unsigned char[::1]`. We previously ran into issues with using const vs non-const in handling numpy arrays that were marked read-only which is what prompted me to questioning an overload.

Reply:
Ok cool.

Yeah I recall some cases where Cython didn't work as well with `const` memoryviews. Issue ( https://github.com/cython/cython/issues/1772 ) comes to mind (though that's with fused types). Not sure if there were issues with `const` memoryviews on a specific type. Was this what you were thinking about?

Reply:
Note: After discussion offline, we decided to add one more test using a read-only NumPy array, which has been included in commit ( https://github.com/rapidsai/rmm/pull/253/commits/2234e5a7c09251375a4586d75e757bcb78f32f59 ).